File: src/backend/FreeBody.java
package backend;

import backend.motion.Motion;

public class FreeBody implements UniversalConstants {

    public Motion motion;
    public double mass;
    public double angularMass;
    public double temperature;
    public double radius;

    public String id = "undefined";

    public FreeBody() {
        this.motion = new Motion();
        this.mass = 1;
        this.angularMass = getAngularMass(1, 1);
        this.temperature = 298.15; // Kelvin
        this.radius = 1;
        this.motion.position.setPosition(250, 250);
    }

    public FreeBody(double mass, double temperature, double radius,
                    double x, double y, double h,
                    double dx, double dy, double dh,
                    String id, int trailersPopulation) {
        this();
        this.mass = mass;
        this.angularMass = getAngularMass(mass, radius);
        this.temperature = temperature;
        this.radius = radius;
        this.motion = new Motion(x, y, h, dx, dy, dh, trailersPopulation);
        this.id = id;
    }

    public double getAngularMass(double mass, double area) {
        return mass * area; // todo: find real equation
    }

}

package backend;

import backend.motion.Acceleration;
import backend.motion.Motion;
import backend.motion.PositionDouble;
import backend.motion.Velocity;

public class FreeBody implements UniversalConstants {

    public Motion motion;
    public double mass;
    public double angularMass;
    public double temperature;
    public double radius;

    public String id = "undefined";

    public FreeBody() {
        this.motion = new Motion();
        this.mass = 1;
        this.angularMass = getAngularMass(1, 1);
        this.temperature = 298.15; // Kelvin
        this.radius = 1;
        this.motion.position.setPosition(250, 250);
    }

    public FreeBody(double mass, double temperature, double radius,
                    double x, double y, double h,
                    double dx, double dy, double dh,
                    String id, int trailersPopulation) {
        this();
        this.mass = mass;
        this.angularMass = getAngularMass(mass, radius);
        this.temperature = temperature;
        this.radius = radius;
        this.motion = new Motion(x, y, h, dx, dy, dh, trailersPopulation);
        this.id = id;
    }

    public double getAngularMass(double mass, double area) {
        return mass * area; // todo: find real equation
    }

    public Acceleration getRelativeAcceleration(FreeBody client) {
        return motion.acceleration.getRelativeAcceleration(client.motion.acceleration);
    }

    public Velocity getRelativeVelocity(FreeBody client) {
        return motion.velocity.getRelativeVelocity(client.motion.velocity);
    }

    public double getDistance(FreeBody client) {
        return motion.position.getDistance(client.motion.position);
    }

    public double getDirection(FreeBody client) {
        return motion.position.getDirection(client.motion.position);
    }

}

File: src/main/kotlin/engine/FreeBody.kt
package engine

import engine.motion.Acceleration
import engine.motion.Location
import engine.motion.Motion
import engine.motion.Velocity
import org.jbox2d.collision.shapes.Shape
import org.jbox2d.dynamics.Body

open class FreeBody(
    val id: String,
    var motion: Motion,
    var shapeBox: Shape?,
    var worldBody: Body?,
    var mass: Double = 1.0,
    var temperature: Double = .0,
    var radius: Double = .0,
    val restitution: Double = 1.0
) {

    val angularMass: Double
        get() = calculateAngularMass()

    constructor(
        id: String = "",
        mass: Double = 1.0,
        temperature: Double = .0,
        radius: Double = .0,
        x: Double = .0, y: Double = .0, h: Double = .0,
        dx: Double = .0, dy: Double = .0, dh: Double = .0,
        restitution: Double = 1.0
    ) : this(
        id,
        Motion(Location(x, y, h), Velocity(dx, dy, dh)),
        null,
        null,
        mass,
        temperature,
        radius,
        restitution
    )

    private fun calculateAngularMass(): Double {
        val area = radius * 4
        return mass * area // TODO: find real equation
    }

    fun getRelativeAcceleration(client: FreeBody): Acceleration {
        return motion.acceleration.getRelative(client.motion.acceleration)
    }

    fun getRelativeVelocity(client: FreeBody): Velocity {
        return motion.velocity.getRelative(client.motion.velocity)
    }

    fun getDistance(client: FreeBody): Double {
        return motion.location.getDistance(client.motion.location)
    }

    fun getDirection(client: FreeBody): Double {
        return motion.location.getDirection(client.motion.location)
    }

}
